File: android/src/main/java/expo/modules/settings/ExpoSettingsModule.kt
package expo.modules.settings

import android.content.Context
import android.content.SharedPreferences
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoSettingsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoSettings")

    Events("onChangeTheme")

    //get SharedPreferences Editor instance and set value
    Function("setTheme") { theme: String ->
      getPreferences().edit().putString("theme", theme.value).commit()
      //bundleOf creates Bundle instances that allow eventually for real-time events subscription
      //from the JS side
      this@ExpoSettingsModule.sendEvent("onChangeTheme", bundleOf("theme" to theme.value))
    }

    //get value from sharedpreferences or return system if nnull
    Function("getTheme") {
      return@Function getPreferences().getString("theme", Theme.SYSTEM.value)
    }
  }

  private val context
    get() = requireNotNull(appContext.reactContext)

  private fun getPreferences(): SharedPreferences {
    return context.getSharedPreferences(context.packageName + ".settings", Context.MODE_PRIVATE)
  }
}

enum class Theme(val value: String): Enumerable {
  LIGHT("light")
  DARK("dark")
  SYSTEM("system")
}

File: ios/ExpoSettingsModule.swift
import ExpoModulesCore

public class ExpoSettingsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoSettings")
      
    //ref: https://docs.expo.dev/modules/module-api/#events
      Events("onChangeTheme")
      
    //set theme value
      Function("setTheme") { (theme: Theme) -> Void in
          UserDefaults.standard.set(theme.rawValue, forKey: "theme")
          //allows for quick sbscriptions to the change events on the JS side from the native side
          sendEvent("onChangeTheme", [
            "theme": theme.rawValue
          ])
      }

    //gets value from UserDefault string or fallback
    Function("getTheme") { () -> String in
        UserDefaults.standard.string(forKey: "theme") ?? Theme.system.rawValue
    }
  }
    
    enum Theme: String, Enumerable {
        case light
        case dark
        case system
    }
}
